[frontend] Fix tensor list alias annotation

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83921

For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

larryliu0820 committed:
[frontend] Fix tensor list alias annotation

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

shunting314 commented on torch/csrc/jit/frontend/function_schema_parser.cpp:
```diff
@@ -165,7 +165,12 @@ struct SchemaParser {
+          auto temp = at::AliasInfo();
+          temp.setIsWrite(alias_info->isWrite());
+          container = c10::optional<at::AliasInfo>(temp);
+        }
```
NIT: 
```
container = c10::optional<at::AliasInfo>(at::AliasInfo());
container->setIsWrite(alias_info->isWrite());
```

It's easier to understand the code with less identifiers.

shunting314 commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -550,15 +550,21 @@ inline std::ostream& operator<<(std::ostream& out, const Argument& arg) {
 
-  if (arg.alias_info()) {
+  // print alias info if it has beforeSets.
+  if (arg.alias_info() && !arg.alias_info()->beforeSets().empty()) {
```
Should we have an API in AliasInfo indicating if the alias info worth printing? We can check if beforeSets is empty inside that API

bdhirsh replied:
Is it possible to have an alias info with nothing in the BeforeSet? Maybe we should just ban that state altogether (assert in the `AliasInfo` constructor that the before set is non-empty)

larryliu0820 replied:
> Is it possible to have an alias info with nothing in the BeforeSet?

Yeah I don't see any use case. Let me add an assert and see what's happen in CI

larryliu0820 replied:
Oh it's an implicit constructor though. I don't see problem of not printing alias info that has an empty before sets, I can add a unit test to assert that it's not printing

bdhirsh replied:
Ah ok! Seems fine then.

shunting314 commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -550,15 +550,21 @@ inline std::ostream& operator<<(std::ostream& out, const Argument& arg) {
-    out << list->getElementType()->str() << "[" << *arg.N() << "]";
+    out << list->getElementType()->str();
+    if (arg.alias_info() && !arg.alias_info()->containedTypes().empty()){
+      out << arg.alias_info()->containedTypes()[0] << "[]";
```
what if arg.N() exist in this case?

shunting314 reviewed (commented):
Leave some comments by literally reading the code. Brian should have more insights :)

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

bdhirsh commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -828,11 +828,15 @@ void AliasDb::analyzeImpl(Node* node) {
     TORCH_INTERNAL_ASSERT(
         !formal->isWildcardBefore(),
         "Doesn't make sense for a input value to begin as a wildcard");
+    if (formal->containedTypes().size() == 1 && formal->beforeSets().empty()) {
```
oh ok, IIUC, this is because `formal->containedTypes().size() == 1 && formal->beforeSets().empty()` is still a valid state that technically means "no alias annotations". Maybe that's worth a comment?

bdhirsh reviewed (approved):
thanks for the fix!

larryliu0820 commented:
@pytorchbot merge -f "All CI except build-docs pass"

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[frontend] Fix tensor list alias annotation"


For issue #77920

The current logic checks alias info before `[]` and after. If no alias info exists after `[]`, we overwrite the alias info before. This logic failed on argument like `Tensor(a!)[]`, dropping the alias info before `[]` on the floor. This PR adds a new alias info if it's missing after `[]`. This way we can keep the alias info before `[]`.

[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge -l

larryliu0820 commented:
@pytorchbot merge -g

larryliu0820 commented:
@pytorchbot merge
